File: src/main/java/com/wowo/wowo/controller/TransactionController.java
package com.wowo.wowo.controller;

import com.wowo.wowo.annotation.authorized.IsUser;
import com.wowo.wowo.data.dto.PagingDTO;
import com.wowo.wowo.data.dto.TransactionDTO;
import com.wowo.wowo.data.dto.TransactionHistoryResponseDTO;
import com.wowo.wowo.service.TransactionService;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Slf4j
@AllArgsConstructor
@RestController
@RequestMapping(value = "v1/transaction", produces = "application/json; charset=UTF-8")
@IsUser
@Tag(name = "Transaction", description = "Giao dịch")
public class TransactionController {

    private final TransactionService transactionService;

    @GetMapping("/{id}")
    public ResponseEntity<TransactionDTO> getTransaction(@PathVariable String id,
            Authentication authentication) {
        log.info("User with id {} is getting transaction with id {}", authentication.getPrincipal(),
                id);
        final TransactionDTO transactionDetail = transactionService.getTransactionDetail(id,
                authentication);
        return ResponseEntity.ok(transactionDetail);
    }

    @GetMapping("/history")
    public TransactionHistoryResponseDTO getAllTransaction(@ModelAttribute @Validated PagingDTO allParams,
            Authentication authentication) {
        return getTransactionHistory(allParams, authentication.getPrincipal()
                .toString());
    }

    private TransactionHistoryResponseDTO getTransactionHistory(PagingDTO allParams,
            String userId) {
        int offset = allParams.getOffset();
        int page = allParams.getPage();
        offset = Math.min(30, Math.max(offset, 0));

        final List<TransactionDTO> recentTransactions = transactionService.getRecentTransactions(
                userId,
                offset,
                page);

        long total = transactionService.getTotalTransactions(
                (userId));
        return new TransactionHistoryResponseDTO(recentTransactions, total);
    }

    //    @IsAdmin
    @GetMapping("/{userId}/history")
    public TransactionHistoryResponseDTO getAllTransactionsByUserId(
            @ModelAttribute @Validated PagingDTO allParams,
            @PathVariable String userId) {

        return getTransactionHistory(allParams, userId);
    }

    @PostMapping("{id}/refund")
    public ResponseEntity<?> refundTransaction(@PathVariable String id) {
        //TODO: Implement refundTransaction
        return ResponseEntity.ok()
                .build();
    }

}

File: src/main/kotlin/com/wowo/wowo/contexts/transaction/presentation/rest/TransactionController.kt
package com.wowo.wowo.contexts.transaction.presentation.rest

import com.wowo.wowo.contexts.transaction.application.dto.TransferMoneyCommand
import com.wowo.wowo.contexts.transaction.application.dto.TransactionDTO
import com.wowo.wowo.contexts.transaction.application.usecase.TransferMoneyUseCase
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

/**
 * REST Controller for Transaction operations
 */
@RestController
@RequestMapping("/api/v1/transactions")
class TransactionController(
    private val transferMoneyUseCase: TransferMoneyUseCase
) {

    @PostMapping("/transfer")
    fun transferMoney(@RequestBody request: TransferMoneyRequest): ResponseEntity<TransactionDTO> {
        val command = TransferMoneyCommand(
            fromWalletId = request.fromWalletId,
            toWalletId = request.toWalletId,
            amount = request.amount,
            currency = request.currency,
            description = request.description
        )

        val transactionDTO = transferMoneyUseCase.execute(command)
        return ResponseEntity.status(HttpStatus.CREATED).body(transactionDTO)
    }
}

data class TransferMoneyRequest(
    val fromWalletId: String,
    val toWalletId: String,
    val amount: String,
    val currency: String,
    val description: String?
)

